### ISSUE TITLE
Static library is actually import library

### ISSUE THREAD

**Issue by jeroen** [qpdf/qpdf#225]
The `libqpdf.a` that is currently in `qpdf-8.1.0-bin-mingw64.zip` and `qpdf-8.1.0-bin-mingw32.zip` is not a static library but an import library. It should be named `libqpdf.dll.a`. When linking against this library, the resulting executable will depend on the `qpdf21.dll` dll file.

I would like to build a standalone executable that does not depend on `qpdf21.dll`. Is it possible to build the actual `libqpdf.a` static library?

**Comment by jberkenbilt** [comment]
I can look into this, but it probably won't happen soon. The first time I tried to do static builds on Windows, things broke because of issues with Windows C++ compilers throwing exceptions across DLL boundaries and how this interacted with third party dependencies. However, that was probably several compiler versions ago, so maybe those issues have been resolved.

Most likely you could make it work by editing make/msvc.mk or make/mingw.mk and changing the flags. Ultimately enabling this will be making those files honor the configure flags you mentioned, assuming I can figure out what the flags should be.

**Comment by jeroen** [comment]
The exceptions shouldn't be an issue. I think all that is needed is fix the makefile to `ar` all the objects together into a `libqpdf.a` and rename the import lib to `libqpdf.dll.a`

**Comment by jberkenbilt** [comment]
> I think all that is needed is fix the makefile to ar all the objects together into a libqpdf.a and rename the import lib to libqpdf.dll.a

Won't that break backward compatibility of people's builds? What if I create a `libqpdf-static.a` instead?

**Comment by jeroen** [comment]
> Won't that break backward compatibility of people's builds? What if I create a libqpdf-static.a instead?

Assuming that they are linking using `-lqpdf`, then the linker will prefer the import library `libqpdf.dll.a` over `libqpdf.a`. So it should not break.

**Comment by jberkenbilt** [comment]
Oh, I didn't know about that. Thanks. I'll aim to fix this for the next update.

**Comment by jeroen** [comment]
From the ld manual: 

```
  [...] when ld is called with the argument `-lxxx' it will attempt to 
  find, in the first directory of its search path, 

            libxxx.dll.a 
            xxx.dll.a 
            libxxx.a 
            xxx.lib 
            cygxxx.dll (*) 
            libxxx.dll 
            xxx.dll 
```

**Comment by jberkenbilt** [comment]
Do you know whether there's any equivalent for msvc? I may just make this work for the mingw builds only.

**Comment by jeroen** [comment]
Equivalent for a static library you mean? Or for the name?

**Comment by jberkenbilt** [comment]
I'm going to close this issue with a commit that renames the library and changes some internal build code to make it clearer that this is an import library, not a static library. If you would like the qpdf distribution to ship with a static library as well, please open a separate ticket. I don't know when/if I will get to doing that, but it's possible that someone may offer a patch. Right now, qpdf's build doesn't explicitly build a static library. On UNIX/Linux systems, we use libtool, which takes care of the chore of building separate static and dynamic libraries. To fix this on Windows, I would have to modify the build myself to build both kinds of libraries, and I don't have time right now to spend on Windows build enhancements. I have one commercial user who rolled their own build which builds static libraries for Windows, so I know that it is definitely doable.

**Comment by jeroen** [comment]
Afaik `mingw-w64` (at least in msys2) uses the same libtool setup as Linux. Hence if `--enable-static` produces the correct static libraries on Linux/Unix, it will probably work for `mingw-w64` as well.

**Comment by jberkenbilt** [comment]
Yes, I use msys2 now for my mingw builds, though when I first coded them, it was before msys2 existed. Probably it's worth checking to see I can make the libtool build code I currenty have work for the mingw builds.
